File: mpi-1x-hypack-2013/point-to-point-mpi-codes/f77-lang-pt2pt-mpi/sum_associative_fanin_nonblocking_tree.f
c
c*******************************************************************
c
c	C-DAC Tech Workshop : hyPACK-2013
c               October 15-18, 2013
c
c  Example 1.10        : sum_associative_fanin_nonblocking_tree.f 
c
c  Objective           : To find sum of 'n' integers on 'p' processors 
c                        using 'Associateive Fan-in' rule.
c                        MPI Non_Blocking Communication library calls are used. 
c
c                        This example demonstrates the use of
c
c                        MPI_Init
c                        MPI_Comm_rank
c                        MPI_Comm_size
c                        MPI_Isend
c                        MPI_Irecv
c                        MPI_Waitall
c                        MPI_Finalize
c
c  Input               : Automatic input data generation
c                        The rank of each process is input on each process.
c
c  Output              : Process with Rank 0 should print the sum of 
c                        'n' values. 
c
c  Necessary Condition : Number of Processes should be 
c                        less than or equal to 8.
c
c   Created           : August-2013
c
c   E-mail            : hpcfte@cdac.in     
c
c*************************************************************************

       program main
       include 'mpif.h'
       integer 	MyRank, Numprocs
       integer  Root
       integer  ilevel, value
       integer  Source, Source_tag
       integer  Destination, Destination_tag
       integer  Level, NextLevel
       real     NoofLevels
       integer  sum, ValidInput
       integer  status(MPI_STATUS_SIZE)
       integer  request(16)

C     ....MPI initialisation....
       call MPI_INIT(ierror)
       call MPI_COMM_SIZE(MPI_COMM_WORLD,Numprocs, ierror)
       call MPI_COMM_RANK(MPI_COMM_WORLD,MyRank, ierror)

C     Check for the no. of processors to be a power of 2 

       Root = 0
       sum = 0
       ValidInput = 0
       do iproc = 1, Numprocs
	  if((2 ** iproc) .eq. Numprocs) then
	    NoofLevels = iproc
	    ValidInput = 1
	    goto 11
	  endif
	enddo
 11	continue

	if(ValidInput .ne. 1) then
   	     if(MyRank .eq. Root) 
     $ 	     print*,"Number of processors should be power of 2"
	     goto 100
	 endif

    	 sum = MyRank 
    	 Source_tag = 0
    	 Destination_tag = 0

    	 do ilevel = 0, NoofLevels-1
	     Level = 2 ** ilevel
	
	     if(mod(MyRank, Level) .eq. 0) then
	           NextLevel = 2 ** (ilevel+1)

C 0123456789012345678901234567890123456789012345678901234567890123456789

	     if(mod(MyRank, NextLevel) .eq. 0) then
	       Source = MyRank + Level
	       call MPI_IRECV(value, 1, MPI_INTEGER, Source, Source_tag, 
     $ 		           MPI_COMM_WORLD, request(ilevel+1), ierror)

	       call MPI_WAIT(request(ilevel+1), status, ierr)
	       sum = sum + value

	     else
	     	        Destination = MyRank - Level
	       call MPI_ISEND(sum, 1, MPI_INTEGER, Destination, 
     $                        Destination_tag, MPI_COMM_WORLD, 
     $                        request(ilevel+1), ierror)
	     endif
	    endif
    	 enddo

	 if(MyRank .eq. Root) print*,'MyRank ',MyRank,'Final SUM ',sum

 100    call MPI_FINALIZE(ierror)

	     stop
	     end

File: mpi-1x-hypack-2013/point-to-point-mpi-codes/c-lang-pt2pt-mpi/sum_associative_fanin_nonblocking_tree.c

/*
***********************************************************************

		C-DAC Tech Workshop : hyPACK-2013
                           October 15-18, 2013

   Example 1.10		: sum_associative_fanin_nonblocking_tree.c

   Objective           : To find sum of 'n' integers on 'p' processors using 
                         'Associative Fan-in' rule.
                         MPI Non-Blocking communication library calls are used.

                         This example demonstrates the use of 
                         MPI_Init
                         MPI_Comm_rank
                         MPI_Comm_size
                         MPI_Isend
                         MPI_Irecv
                         MPI_Waitall
                         MPI_Finalize

   Input               : Automatic input generation
                         The rank of each process is input on each process.

   Output              : Process with Rank 0 should print the sum of 'n' values 

   Necessary           : Number of Processes should be less than 
   Condtion              or equal to 8.

   Created             : August-2013

   E-mail              : hpcfte@cdac.in     


************************************************************************
*/



#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "mpi.h"

#define  LOG2(x)    log10((double)(x)) / log10((double) 2)
#define  IS_INT(x)  ((x) == (int)(x))?1:0

int main(int argc,char *argv[])
{
    int	   MyRank, Numprocs;
    int    Root = 0;
    int    ilevel, value, ans;
    int	   Source, Source_tag;
    int    Destination, Destination_tag;
    int	   Level, NextLevel;
    float  NoofLevels;
    static int	 sum = 0;
    MPI_Status  *status;
    MPI_Request request[200];


    MPI_Init(&argc,&argv);
    MPI_Comm_size(MPI_COMM_WORLD,&Numprocs);
    MPI_Comm_rank(MPI_COMM_WORLD,&MyRank);

    NoofLevels = LOG2(Numprocs);
    if(!(IS_INT(NoofLevels))){
        if(MyRank == Root)
	   printf("\nNumber of processors should be power of 2\n");
        MPI_Finalize();
        exit(-1);
    }

    sum = MyRank; 
    Source_tag = 0;
    Destination_tag = 0; 

    for(ilevel = 0 ; ilevel  < NoofLevels; ilevel ++){
	Level = (int)(pow((double)2, (double)ilevel ));
	
	if((MyRank % Level) == 0){
	    NextLevel = (int)(pow((double)2, (double)(ilevel +1)));

	    if((MyRank % NextLevel) == 0){
	        Source = MyRank + Level;
	        MPI_Irecv(&value, 1, MPI_INT, Source, Source_tag, 
			   MPI_COMM_WORLD, &request[ilevel]);
  	        MPI_Waitall(1,&request[ilevel],status); 
                sum = sum + value;
	    }
	    else{
	        Destination = (MyRank - Level);
	        MPI_Isend(&sum, 1, MPI_INT, Destination, Destination_tag, 
			   MPI_COMM_WORLD,&request[ilevel]);
  	     	MPI_Waitall(1,&request[ilevel],status); 
	    }
        }
    }
	     
    if(MyRank == Root) 
       printf(" My Rank is %d Final SUM is%d\n", MyRank, sum);

    MPI_Finalize();
}



